osutil/disks/disks.go

```diff
@@ -173,6 +195,89 @@ func diskFromMountPointImpl(mountpoint string, opts *Options) (*disk, error) {
+		if props["DEVTYPE"] != "disk" {
+			// not a decrypted device
+			return nil, fmt.Errorf("mountpoint source %s is not a decrypted device: devtype is not disk (is %s)", partMountPointSource, props["DEVTYPE"])
+		}
+
+		// TODO:UC20: currently, we effectively parse the DM_UUID env variable
+		//            that is set for the mapper device volume, but doing so is
+		//            actually wrong, since the value of DM_UUID is an
+		//            implementation detail that depends on the subsystem
+		//            "owner" of the device such that the prefix is considered
+		//            the owner and the suffix is private data owned by the
+		//            subsystem. In our case, in UC20 initramfs, we have the
+		//            device "owned" by systemd-cryptsetup, so we should ideally
+		//            parse that the first part of DM_UUID matches CRYPT- and
+		//            then use `cryptsetup status` (since CRYPT indicates it is
+		//            "owned" by cryptsetup) to get more information on the
+		//            device sufficient for our purposes to find the encrypted
+		//            device/partition underneath the mapper.
+		//            However we don't currently have cryptsetup in the initrd,
+		//            so we can't do that yet :-(
+
+		// TODO:UC20: these files are also likely readable through udev env
+		//            properties, but it's unclear if reading there is reliable
+		//            or not, given that these variables have been observed to
+		//            be missing from the initrd previously, and are not
+		//            available at all during userspace on UC20 for some reason
+		dmUUID, err := ioutil.ReadFile(filepath.Join(devBlockDir, d.Dev(), "dm", "uuid"))
+		if err != nil && os.IsNotExist(err) {
+			return nil, fmt.Errorf("mountpoint source %s is not a decrypted device: missing device mapper metadata: no dm-uuid", partMountPointSource)
+		}
+
+		dmName, err := ioutil.ReadFile(filepath.Join(devBlockDir, d.Dev(), "dm", "name"))
+		if err != nil && os.IsNotExist(err) {
+			return nil, fmt.Errorf("mountpoint source %s is not a decrypted device: missing device mapper metadata: no dm-name", partMountPointSource)
+		}
+
+		// trim the suffix of the dm name from the dm uuid to safely match the
+		// regex - the dm uuid contains the dm name, and the dm name is user
+		// controlled, so we want to remove that and just use the luks pattern
+		// to match the device uuid
+		// we are extra safe here since the dm name could be hypothetically user
+		// controlled via an external USB disk with LVM partition names, etc.
+		dmUUIDSafe := bytes.TrimSuffix(
+			bytes.TrimSpace(dmUUID),
+			append([]byte("-"), bytes.TrimSpace(dmName)...),
+		)
+		matches := luksUUIDPatternRe.FindSubmatch(dmUUIDSafe)
+		if len(matches) != 2 {
+			// the format of the uuid is different - different luks version
+			// maybe?
+			return nil, fmt.Errorf("cannot verify disk: partition %s does not have a valid luks uuid format", d.Dev())
```

Comment:
Should we output the dmUUIDSafe value here too in the error message? So that when we see the error we get ideas what we need to fix in our code/regex?